pkg/services/ngalert/notifier/silence_svc.go

```diff
@@ -125,3 +143,83 @@ func (s *SilenceService) DeleteSilence(ctx context.Context, user identity.Reques
 
 	return nil
 }
+
+// WithAccessControlMetadata adds access control metadata to the given SilenceWithMetadata.
+func (s *SilenceService) WithAccessControlMetadata(ctx context.Context, user identity.Requester, silencesWithMetadata ...*models.SilenceWithMetadata) error {
+	silences := make([]*models.Silence, 0, len(silencesWithMetadata))
+	for _, silence := range silencesWithMetadata {
+		silences = append(silences, silence.Silence)
+	}
+	permissions, err := s.authz.SilenceAccess(ctx, user, silences)
+	if err != nil {
+		return err
+	}
+
+	if len(permissions) != len(silences) {
+		s.log.Error("failed to get metadata for all silences")
+	}
+
+	for _, silence := range silencesWithMetadata {
+		if perms, ok := permissions[silence.Silence]; ok {
+			silence.Metadata.Permissions = &perms
+		}
+	}
+
+	return nil
+}
+
+// WithRuleMetadata adds rule metadata to the given SilenceWithMetadata.
+func (s *SilenceService) WithRuleMetadata(ctx context.Context, user identity.Requester, silences ...*models.SilenceWithMetadata) error {
+	byRuleUID := make(map[string][]*models.SilenceWithMetadata, len(silences))
+	for _, silence := range silences {
+		ruleUID := silence.GetRuleUID()
+		if ruleUID != nil {
+			byRuleUID[*ruleUID] = append(byRuleUID[*ruleUID], silence)
+		}
+	}
+
+	if len(byRuleUID) == 0 {
+		return nil
+	}
+
+	q := models.ListAlertRulesQuery{
+		RuleUIDs: maps.Keys(byRuleUID),
+		OrgID:    user.GetOrgID(),
+	}
+
+	rules, err := s.ruleStore.ListAlertRules(ctx, &q)
+	if err != nil {
+		return err
+	}
+
+	accessCacheByFolder := make(map[string]bool)
+	for _, rule := range rules {
+		// TODO: Preferably silence service should not need to know about the internal details of rule access control.
+		// This can be improved by adding a method to ruleAuthz that does the filtering itself or a method that exposes
+		// an access fingerprint for a rule that callers can use to do their own caching.
+		fp := rule.NamespaceUID
+		if canAccess, ok := accessCacheByFolder[fp]; ok && !canAccess {
+			continue
+		}
+
+		canAccess, err := s.ruleAuthz.HasAccessInFolder(ctx, user, rule)
+		if err == nil {
+			accessCacheByFolder[fp] = canAccess // Only cache if there is no error.
+		}
+		if err != nil || !canAccess {
+			continue // Assume no access if there is an error.
+		}
+
+		if ruleSilences, ok := byRuleUID[rule.UID]; ok {
```

Comment:
In the case when silence contains ruleUID that does not exist, we do not return any metadata. However, I think it would be better to return "partial" metadata and populate only rule UID. 
This way, UI will be able to detect it as kinda orphaned silence and notify user. WDYT?